### PR TITLE
[inference] Dynamic Batching for Single and Multiple GPUs

### PR DESCRIPTION
**Pull request by CjhHa1**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by tiandiao123 — APPROVED** [review]
temporarily approved. testing rayserve!

File: colossalai/inference/dynamic_batching/infer_batch.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        input_lengths = []
+        all_input_ids = []
+        requests_idx_mapping = {}
+
+        out_token_id_counts = []
+        sampling_param_list = []
+
+        nopad_total_token_num = 0
+        nopad_max_len_in_batch = 0
+        nopad_b_loc = torch.empty((len(requests), max_total_len + 12), dtype=torch.long, device="cuda")
```
Comment (yuehuayingxueluo) [inline colossalai/inference/dynamic_batching/infer_batch.py:78]: Why is 12 added here?

File: colossalai/inference/manager.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+    def _merge_batch(self, batch1, batch2):
+        """
+        Merge new mini batch into running batch.
+        """
+        batch1 = self.engine.cache.pop(batch1.batch_id)
+        batch2 = self.engine.cache.pop(batch2.batch_id)
+
+        m_batch = InferBatch.merge(batch1, batch2)
+        self.engine.cache[batch1.batch_id] = m_batch
+        print("merged_batch", m_batch)
```
Comment (yuehuayingxueluo) [inline colossalai/inference/manager.py:190]: delete it.

File: colossalai/inference/manager.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            log_stats=not args.disable_log_stats,
+            log_stats_interval=args.log_stats_interval,
+            waiting_req_list=waiting_req_list,
+        )
+
+    except Exception:
+        batch_manager.clean_up()
+        raise
+
+    print("start dynamic batching process")
```
Comment (yuehuayingxueluo) [inline colossalai/inference/manager.py:242]: delete it.

File: colossalai/inference/dynamic_batching/infer_batch.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/infer_batch.py:78]: This arg is inherited from light-llm

File: colossalai/inference/manager.py (hunk 2 above)
Comment (CjhHa1) [reply colossalai/inference/manager.py:190]: done

File: colossalai/inference/manager.py (hunk 3 above)
Comment (CjhHa1) [reply colossalai/inference/manager.py:242]: done

File: colossalai/inference/dynamic_batching/infer_batch.py (hunk 1 above)
Comment (ver217) [reply colossalai/inference/dynamic_batching/infer_batch.py:78]: You'd better add comments to claim this

File: colossalai/inference/dynamic_batching/io_struct.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        if self.output_metadata_list:
+            out.gen_metadata.update(self.output_metadata_list[-1])
+        return out
+
+    def stop_sequences_matched(self):
+        # should we add stpp sequences to the sample params?
+        if self.sample_params.stop_sequences is not None:
+            for stop_token_ids in self.sample_params.stop_sequences:
+                stop_len = len(stop_token_ids)
+                if stop_len > 0:
```
Comment (ver217) [inline colossalai/inference/dynamic_batching/io_struct.py:39]:
```python
if a and b and c:
    return True
```

File: colossalai/inference/dynamic_batching/infer_batch.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                    (self.nopad_max_len_in_batch - 1)
+                    - (self.nopad_b_seq_len[idx] - 1) : (self.nopad_max_len_in_batch - 1),
+                ]
+            )
+        remove_index = torch.cat(remove_index, dim=-1)
+        self.cache_manager.free(remove_index)
+        return
+
+    @torch.no_grad()
+    def filter(self, request_ids: List[int]):
```
Comment (ver217) [inline colossalai/inference/dynamic_batching/infer_batch.py:144]: Add type hint for return value

Comment (ver217) [reply colossalai/inference/dynamic_batching/infer_batch.py:144]: For all functions

File: colossalai/inference/dynamic_batching/req_queue.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+        left_out_len_array = np.array([e[1] for e in self.cache_len_list])
+        # assert left_out_len_array.min() >= 0
+        has_run_len_array = np.array([e[0] for e in self.cache_len_list])
+        cum_run_len_array = np.cumsum(has_run_len_array)
+        size_array = np.arange(1, len(self.cache_len_list) + 1, 1)
+
+        need_max_token_num = (left_out_len_array * size_array + cum_run_len_array).max()
+        # NOTE: change here < to <=
+        if need_max_token_num <= self.max_total_tokens and len(self.cache_len_list) <= self.running_max_req_size:
```
Comment (ver217) [inline colossalai/inference/dynamic_batching/req_queue.py:43]: `return a and b `

File: colossalai/inference/dynamic_batching/sampling_params.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        self,
+        do_sample: bool = False,
+        presence_penalty: float = 0.0,
+        frequency_penalty: float = 0.0,
+        temperature: float = 1.0,
+        top_p: float = 1.0,
+        top_k: int = -1,  # -1 is for all 
+        ignore_eos: bool = False,
+        max_new_tokens: int = 16,
+        stop_sequences: Optional[Union[str, List[str]]] = None  # 停止句子条件
```
Comment (ver217) [inline colossalai/inference/dynamic_batching/sampling_params.py:19]: Don't use Chinese

File: colossalai/inference/dynamic_batching/stas.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import time
+
+
+class Stats:
```
Comment (ver217) [inline colossalai/inference/dynamic_batching/stas.py:4]: The filename would be `stats.py`, right?

File: colossalai/inference/manager.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    def loop_for_fwd(self):
+        """
+        The main loop for a dynamic batching process.
+        """
+        counter_count = 0
+        while self.running_batch is not None or self.req_queue.waiting_req_list:
+            self._step()
+            counter_count += 1
+            if self.running_batch is not None:
+                if counter_count % 50 == 0:
```
Comment (ver217) [inline colossalai/inference/manager.py:77]: Make this configurable

File: tests/kit/model_zoo/transformers/llama.py [hunk 10]
```diff
@@ -47,10 +49,10 @@ def data_gen_for_casual_lm():
     loss_fn_for_seq_classification = lambda output: output.logits.mean()
 
     config = LlamaConfig(
-        num_hidden_layers=4,
-        hidden_size=128,
+        num_hidden_layers=8,
```
Comment (ver217) [inline tests/kit/model_zoo/transformers/llama.py:52]: Why use such a large model?

File: tests/test_infer/test_dynamic_batching/test_forward.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+TP_SIZE = 1
+MAX_BATCH_SIZE = 2
+MAX_INPUT_LEN = 5
+MAX_OUTPUT_LEN = 16
+CUDA_SUPPORT = version.parse(torch.version.cuda) > version.parse("11.5")
+
+
+def run():
+    parser = argparse.ArgumentParser()
```
Comment (ver217) [inline tests/test_infer/test_dynamic_batching/test_forward.py:25]: This is useless

File: colossalai/inference/dynamic_batching/infer_batch.py (hunk 1 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/infer_batch.py:78]: added

File: colossalai/inference/dynamic_batching/infer_batch.py (hunk 5 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/infer_batch.py]: added!

File: colossalai/inference/dynamic_batching/req_queue.py (hunk 6 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/req_queue.py]: done

File: colossalai/inference/dynamic_batching/sampling_params.py (hunk 7 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/sampling_params.py]: updated

File: colossalai/inference/dynamic_batching/stas.py (hunk 8 above)
Comment (CjhHa1) [reply colossalai/inference/dynamic_batching/stas.py]: for sure

File: colossalai/inference/manager.py (hunk 9 above)
Comment (CjhHa1) [reply colossalai/inference/manager.py]: updated

File: tests/kit/model_zoo/transformers/llama.py (hunk 10 above)
Comment (CjhHa1) [reply tests/kit/model_zoo/transformers/llama.py]: it was a mistake due to automatic merge, fixed

File: tests/test_infer/test_dynamic_batching/test_forward.py (hunk 11 above)
Comment (CjhHa1) [reply tests/test_infer/test_dynamic_batching/test_forward.py]: updated

**Review by ver217 — APPROVED** [review]

**Pull request merged**